### ISSUE TITLE
[Bug]: unable to import zod-form-data with Vite SSR, package is treated as a CJS module

### ISSUE THREAD

**Issue by geoffrich** [airjp73/remix-validated-form#230]
### Which packages are impacted?

- [ ] `remix-validated-form`
- [ ] `@remix-validated-form/with-zod`
- [ ] `@remix-validated-form/with-yup`
- [X] `zod-form-data`

### What version of these packages are you using?

`zod-form-data`: 1.2.4

### Please provide a link to a minimal reproduction of the issue.

https://github.com/geoffrich/vite-zfd-repro

### Steps to Reproduce the Bug or Issue

1. Create a new Vite SSR project with `npm create vite-extra@latest repro-project -- --template ssr-vanilla`
2. Install `zod` and `zod-form-data`
3. Add the following to `/src/entry-server.js`
```js
import { zfd } from "zod-form-data";

const schema = zfd.formData({
  name: zfd.text()
});

export function render() {
  const params = schema.parse(new URLSearchParams("name=test"));

  const html = `
    ${JSON.stringify(params)}
  `;
  return { html };
}
```
4. Run `npm run build` and `npm run preview`. See the following error. The issue does not manifest in dev mode, since Vite bundles the package differently.

```
file:///Users/grich/code/other/vite-zfd-repro/dist/server/entry-server.js:1
import { zfd } from "zod-form-data";
         ^^^
SyntaxError: Named export 'zfd' not found. The requested module 'zod-form-data' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'zod-form-data';
const { zfd } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:128:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:194:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:385:24)
    at async file:///Users/grich/code/other/vite-zfd-repro/server.js:51:17
```

### Expected behavior

I expected there not to be an error. 

I can workaround it by adding zod-form-data to [ssr.noExternal](https://vitejs.dev/config/ssr-options.html#ssr-noexternal) in the Vite config to force it to be bundled, but this error indicates a packaging issue with the package and would be good to have fixed.

### Platform

- OS: MacOS
- Node version: 16.16.0
- Browser: n/a

### Additional context

I first ran into this when trying to use this package with SvelteKit, though the provided repo only uses Vite to make reproduction simpler.

I was able to fix it by 1) renaming `/dist/zod-form-data.es.js` to `/dist/zod-form-data.es.mjs` and 2) adding `"exports": "./dist/zod-form-data.es.mjs"` to the `package.json`. This was suggested by [Publint](https://publint.dev/zod-form-data@1.2.4), which validates `package.json` files to make sure they are valid modules.

Not sure if that has any unintended consequences, but it seemed to do the trick locally.

Thanks for this package, it's been great to use to validate query params!

**Comment by kyuuaria** [comment]
Any news on this?

**Comment by airjp73** [comment]
What publint is suggesting sounds reasonable to me. Would you mind opening a PR?

**Comment by kyuuaria** [comment]
I could open a PR too if you are fine with that

**Comment by einarpersson** [comment]
Also experience this. Thanks @geoffrich for temporary workaround 🙏

**Comment by multiplehats** [comment]
Same issue here!

**Comment by kyuuaria** [comment]
Well it's up to @airjp73 how to further proceed with my pr

**Comment by CapitaineToinon** [comment]
Adding an alias into your vite config can fix the problem as a workaround. 

```ts
import { sveltekit } from '@sveltejs/kit/vite'
import type { UserConfig } from 'vite'

const config: UserConfig = {
	plugins: [sveltekit()],
	test: {
		include: ['src/**/*.{test,spec}.{js,ts}']
	},
	resolve: {
		alias: {
			'zod-form-data': './node_modules/zod-form-data/dist/zod-form-data.es.js'
		}
	}
}

export default config
```

**Comment by kyuuaria** [comment]
The underlying issue is that the packages have been interpreted as CJS so far so I unfortunately don't see a way to fix this without a breaking change for remix users

**Comment by airjp73** [comment]
Could you all try `zod-form-data@1.2.5-beta.0` to see if that fixes it?

**Comment by kyuuaria** [comment]
> Could you all try `zod-form-data@1.2.5-beta.0` to see if that fixes it?

It actually works

**Comment by CapitaineToinon** [comment]
Seems to be working for the module part but it broke my types. I was probably tying my functions incorrectly to beging with though, currently typing stuff created using `zfd.formData` as `z.ZodEffects<z.AnyZodObject>`.

**Comment by airjp73** [comment]
Published as v1.3.0. Turns out the existing build configuration works and the version on npm is outdated.

**Comment by kyuuaria** [comment]
> Seems to be working for the module part but it broke my types. I was probably tying my functions incorrectly to beging with though, currently typing stuff created using `zfd.formData` as `z.ZodEffects<z.AnyZodObject>`.

Can confirm it broke the types

**Comment by airjp73** [comment]
> Seems to be working for the module part but it broke my types. I was probably tying my functions incorrectly to beging with though, currently typing stuff created using `zfd.formData` as `z.ZodEffects<z.AnyZodObject>`.

Can you provide an example of what's broken? Are the types completely missing or just different?

**Comment by CapitaineToinon** [comment]
`"zod-form-data": "^1.2.4"`

```ts
import { z } from 'zod'

export const login = z.object({
	email: z.string().email()
})

const loginForm = zfd.formData(zodLogin)
```

With this, loginForm was typed like this
```ts
const loginForm: ZodEffects<ZodObject<{
    email: ZodString;
}, "strip", ZodTypeAny, {
    email: string;
}, {
    email: string;
}>, {
    email: string;
}, {
    email: string;
}>
```

so I could have a function expecting that zfd schema like this

```ts
export async function parseForm<T extends z.AnyZodObject>({
	schema,
	formData
}: {
	schema: z.ZodEffects<T>
	formData: FormData
}) {

// ...

}
```

Now it's typed like this 

```ts
const loginForm: ZodEffects<ZodObject<{
    email: ZodString;
}, "strip", ZodTypeAny, {
    email: string;
}, {
    email: string;
}>, {
    email: string;
}, unknown>
```

which now throws when trying to pass it to the parseForm function.

```
Type 'ZodEffects<ZodObject<{ email: ZodString; }, "strip", ZodTypeAny, { email: string; }, { email: string; }>, { email: string; }, unknown>' is not assignable to type 'ZodEffects<ZodObject<{ email: ZodString; }, "strip", ZodTypeAny, { email: string; }, { email: string; }>, { email: string; }, { email: string; }>'.
  Type 'unknown' is not assignable to type '{ email: string; }'.ts(2322)
form.ts(47, 2): The expected type comes from property 'schema' which is declared here on type '{ schema: ZodEffects<ZodObject<{ email: ZodString; }, "strip", ZodTypeAny, { email: string; }, { email: string; }>, { email: string; }, { email: string; }>; formData: FormData; }'
```

That last param being now unknown seems to cause issues.

Again though, if there is a better way to expect a zfd.formData object I'm all up for it cause right now I figured out the signature myself and it worked so far.

**Comment by airjp73** [comment]
Found the issue there. Will put out a fix soon.

**Comment by airjp73** [comment]
Released in v1.3.1

**Comment by CapitaineToinon** [comment]
Now it seems to work in dev but breaks in production when calling `safeParseAsync`. Wonder if we're back at the bundling problem

```
[WebServer] TypeError: keyValidator._parse is not a function
    at ZodObject._parse (/home/runner/work/leaderboards/leaderboards/node_modules/zod/lib/types.js:[14](https://github.com/CapitaineToinon/leaderboards/actions/runs/4128249678/jobs/7132456395#step:9:15)25:37)
    at ZodObject._parseAsync (/home/runner/work/leaderboards/leaderboards/node_modules/zod/lib/types.js:118:29)
    at /home/runner/work/leaderboards/leaderboards/node_modules/zod/lib/types.js:2587:45
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async ZodEffects.safeParseAsync (/home/runner/work/leaderboards/leaderboards/node_modules/zod/lib/types.js:[16](https://github.com/CapitaineToinon/leaderboards/actions/runs/4128249678/jobs/7132456395#step:9:17)4:24)
    at async parseForm (file:///home/runner/work/leaderboards/leaderboards/.svelte-kit/output/server/chunks/form.js:111:[18](https://github.com/CapitaineToinon/leaderboards/actions/runs/4128249678/jobs/7132456395#step:9:19))
    at async default (file:///home/runner/work/leaderboards/leaderboards/.svelte-kit/output/server/entries/pages/auth/login/_page.server.ts.js:16:18)
    at async handle_action_json_request (file:///home/runner/work/leaderboards/leaderboards/.svelte-kit/output/server/index.js:316:18)
    at async resolve (file:///home/runner/work/leaderboards/leaderboards/.svelte-kit/output/server/index.js:2365:[22](https://github.com/CapitaineToinon/leaderboards/actions/runs/4128249678/jobs/7132456395#step:9:23))
    at async Object.handle (file:///home/runner/work/leaderboards/leaderboards/.svelte-kit/output/server/chunks/hooks.server.js:43:10)
```

That's the output of playwright, same test used to work on previous version

**Comment by CapitaineToinon** [comment]
Adding an alias again with 1.3.1 works again. So maybe still a module resolution problem?

```ts
import type { UserConfig } from 'vite'
import { sveltekit } from '@sveltejs/kit/vite'
import { resolve } from 'path'

const projectRootDir = resolve(__dirname)

const config: UserConfig = {
	plugins: [sveltekit()],
	test: {
		include: ['src/**/*.{test,spec}.{js,ts}']
	},
	resolve: {
		alias: [
			{
				find: 'zod-form-data',
				replacement: resolve(projectRootDir, 'node_modules/zod-form-data/dist/index.esm.js')
			}
		]
	}
}

export default config
```

**Comment by kyuuaria** [comment]
I'm also using sveltekit and parsing with safeParseAsync and it works for me in both dev and prod

**Comment by CapitaineToinon** [comment]
> I'm also using sveltekit and parsing with safeParseAsync and it works for me in both dev and prod

Interesting! Do you mind sharing an example schema you're calling safeParseAsync on? Also what version of zod are you on? I'm on 3.20.5

**Comment by kyuuaria** [comment]
Something that seems odd to me is the increased [bundle size](https://bundlephobia.com/package/zod-form-data@1.3.1)

**Comment by airjp73** [comment]
> Something that seems odd to me is the increased [bundle size](https://bundlephobia.com/package/zod-form-data@1.3.1)

The extra bundle size is because we rely on one lodash function ☹️. I wrote a custom replacement for `get` and `set` but I had to roll back the `get` replacement because there are some edge-cases it handled differently.

---

It does seem like this new error is a bundling issue again. Somehow we're ending up with two instances of zod. So when do [`shapeOrSchema instanceof ZodType`](https://github.com/airjp73/remix-validated-form/blob/main/packages/zod-form-data/src/helpers.ts#L178), it thinks its a plain object and passes the schema to `z.object` an additional time.

**Comment by airjp73** [comment]
Okay, let's give this one more go haha. `1.3.2-beta.0` should be fixed.

**Comment by CapitaineToinon** [comment]
1.3.2-beta.0 seems to fix my issues! For the bundle problem, have you tried https://www.npmjs.com/package/lodash-es as a drop-in replacement?

**Comment by multiplehats** [comment]
Can confirm that the version before `1.3.2-beta.0` broke for me in SvelteKit. But `1.3.2-beta.0` works nicely. 

Thanks for all your hard work 👏

**Comment by airjp73** [comment]
The fix has been officially published in zfd 2.0.0.

It's a major version because the types were changed in #243. These types are more accurate, but it's breaking if you've been relying on the exact return type of `formData`. Where possible, it's a good idea to inver the type of the schema, rather than trying to exactly write out the expect type yourself.
